File: moderate/lost_in_translation.c
#include <stdbool.h>
#include <stdio.h>

static const char *codel = "nug"
	"rbc vjnmkf kd yxyqci na rbc zjkfoscdd ew rbc ujllmcp"
	"tc rbkso rbyr ejp mysljylc kd kxveddknmc re jsicpdrysi"
	"de kr kd eoya kw aej icfkici re zjkr";
static const char *decol = "bjv"
	"the public is amazed by the quickness of the juggler"
	"we think that our language is impossible to understand"
	"so it is okay if you decided to quit";

int main(int argc, char *argv[]) {
	FILE *fp;
	bool p = true;
	char c;
	char m[26] = { 0 }, n[26] = { 0 };
	unsigned i = 0, j;

	while (codel[i] != '\0') {
		if (codel[i] >= 'a' && codel[i] <= 'z') {
			m[codel[i] - 'a'] = decol[i];
			n[decol[i] - 'a'] = codel[i];
		}
		i++;
	}
	for (i = 0; i < 26; i++) {
		if (m[i] == 0) {
			for (j = 0; j < 26; j++) {
				if (n[j] == 0) {
					m[i] = (char)j + 'a';
					n[j] = (char)i + 'a';
					break;
				}
			}
		}
	}

	if (argc != 2) {
		printf("Usage: %s [FILE]\n", argv[0]);
		return 1;
	}
	fp = fopen(*++argv, "r");
	while ((c = getc(fp)) != EOF) {
		if (c >= 'a' && c <= 'z')
			putchar(m[c - 'a']);
		else
			putchar(c);
		p = c == '\n';
	}
	if (!p)
		putchar('\n');
	return 0;
}

File: moderate/lost_in_translation.hs
import System.Environment (getArgs)
import Data.List (elem)
import Data.Map (Map)
import qualified Data.Map as Map

codel :: String
codel  = "nug\
    \ rbc vjnmkf kd yxyqci na rbc zjkfoscdd ew rbc ujllmcp\
    \ tc rbkso rbyr ejp mysljylc kd kxveddknmc re jsicpdrysi\
    \ de kr kd eoya kw aej icfkici re zjkr"
decol :: String
decol  = "bjv\
    \ the public is amazed by the quickness of the juggler\
    \ we think that our language is impossible to understand\
    \ so it is okay if you decided to quit"

codeMap                    :: String -> String -> String -> String -> Map Char Char
codeMap (x:xs) (y:ys) cs ds | elem x cs = codeMap xs (y:ys) cs ds
                            | elem y ds = codeMap (x:xs) ys cs ds
                            | otherwise = Map.insert x y (Map.fromList (zip cs ds))

lostInTranslation       :: String -> String
lostInTranslation []     = []
lostInTranslation (x:xs) = Map.findWithDefault ' ' x cMap : lostInTranslation xs
                         where cMap = codeMap ['a'..'z'] ['a'..'z'] codel decol

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map lostInTranslation $ lines input
